File: dashboard/scheduler.py
"""The background loop: due schedules, idle machines, nightly housekeeping."""

import time
from datetime import datetime, timedelta, timezone

import audit
import backups
import fleet
import shares
from db import connect
from machines import list_computers, now_iso
from settings import IDLE_SUSPEND_MINUTES
from tasks import dispatch_task


def compute_next_run(kind: str, every_minutes: int | None, at_time: str | None,
                     after: datetime | None = None) -> str:
    now = after or datetime.now(timezone.utc)
    if kind == "interval":
        return (now + timedelta(minutes=every_minutes or 60)).isoformat(timespec="seconds")
    hh, mm = (at_time or "09:00").split(":")
    nxt = now.replace(hour=int(hh), minute=int(mm), second=0, microsecond=0)
    if nxt <= now:
        nxt += timedelta(days=1)
    return nxt.isoformat(timespec="seconds")


def scheduler_tick() -> None:
    """Dispatch every schedule that has come due.

    Claiming is a conditional UPDATE on next_run_at: with several gunicorn
    workers each running this loop, only the one whose UPDATE matches the row
    it read gets to fire, so a schedule never double-dispatches.
    """
    now = datetime.now(timezone.utc)
    conn = connect()
    due = conn.execute(
        "SELECT * FROM schedules WHERE enabled = 1 AND next_run_at <= ?",
        (now.isoformat(timespec="seconds"),),
    ).fetchall()

    for row in due:
        nxt = compute_next_run(row["kind"], row["every_minutes"], row["at_time"], after=now)
        claimed = conn.execute(
            "UPDATE schedules SET next_run_at = ?, last_run_at = ?, run_count = run_count + 1 "
            "WHERE id = ? AND next_run_at = ?",
            (nxt, now.isoformat(timespec="seconds"), row["id"], row["next_run_at"]),
        )
        conn.commit()
        if claimed.rowcount != 1:
            continue  # another worker got it first
        try:
            dispatch_task(row["desktop"], row["description"])
        except Exception:  # noqa: BLE001
            pass
    conn.close()


def idle_tick() -> None:
    """Put machines nobody is watching to sleep.

    Two things count as "in use": a browser with the screen open (an
    established connection to websockify inside the desktop) and a task that
    is pending or running. Anything else has been idle since last_active_at,
    and once that is older than the timeout the machine is stopped.

    Deliberately off by default. Sleeping frees the machine's memory but ends
    its X session, so a surprise suspend costs someone their open windows —
    that has to be a choice, not a default. Machines flagged no_suspend are
    always skipped.
    """
    if IDLE_SUSPEND_MINUTES <= 0:
        return
    cutoff = datetime.now(timezone.utc) - timedelta(minutes=IDLE_SUSPEND_MINUTES)
    conn = connect()
    busy = {r["desktop"] for r in conn.execute(
        "SELECT DISTINCT desktop FROM tasks WHERE status IN ('PENDING','RUNNING')")}
    rows = [dict(r) for r in conn.execute(
        "SELECT * FROM computers WHERE no_suspend = 0")]

    for comp in rows:
        if comp["name"] in busy:
            conn.execute("UPDATE computers SET last_active_at = ? WHERE id = ?",
                         (now_iso(), comp["id"]))
            continue
        watchers = fleet.vnc_watchers(comp["slug"])
        if watchers is None:
            continue                      # already asleep, or unreachable
        if watchers > 0:
            conn.execute("UPDATE computers SET last_active_at = ? WHERE id = ?",
                         (now_iso(), comp["id"]))
            continue
        last = comp["last_active_at"]
        if not last:
            # Never seen active — start the clock now rather than suspending a
            # machine the moment the feature is switched on.
            conn.execute("UPDATE computers SET last_active_at = ? WHERE id = ?",
                         (now_iso(), comp["id"]))
            continue
        try:
            if datetime.fromisoformat(last) > cutoff:
                continue
        except ValueError:
            continue
        try:
            fleet.suspend_computer(comp["slug"])
        except Exception:  # noqa: BLE001
            pass
    conn.commit()
    conn.close()


def claim_daily(key: str, at_time: str) -> bool:
    """True at most once per UTC day, for whichever worker gets there first.

    Same conditional-UPDATE trick the task scheduler uses: several gunicorn
    workers run this loop, and a fleet backup that fires once per worker would
    be three times the disk and three times the wall clock.
    """
    try:
        hh, mm = [int(x) for x in at_time.split(":", 1)]
    except (ValueError, AttributeError):
        return False
    now = datetime.now(timezone.utc)
    if (now.hour, now.minute) < (hh, mm):
        return False
    today = now.date().isoformat()

    conn = connect()
    try:
        row = conn.execute("SELECT value FROM meta WHERE key = ?", (key,)).fetchone()
        if row and row["value"] >= today:
            return False
        if row:
            claimed = conn.execute(
                "UPDATE meta SET value = ? WHERE key = ? AND value = ?",
                (today, key, row["value"]))
        else:
            claimed = conn.execute(
                "INSERT OR IGNORE INTO meta (key, value) VALUES (?, ?)", (key, today))
        conn.commit()
        return claimed.rowcount == 1
    finally:
        conn.close()


def maintenance_tick() -> None:
    """Daily housekeeping: back the fleet up, then trim what has aged out."""
    if backups.DAILY_AT and claim_daily("backup_daily", backups.DAILY_AT):
        for comp in list_computers():
            try:
                meta = backups.create(comp["slug"])
                audit.record("backup (scheduled)", target=comp["name"],
                             detail=f"{meta['name']} ({meta['bytes']} bytes)")
            except Exception as exc:  # noqa: BLE001
                audit.record("backup (scheduled)", target=comp["name"],
                             detail=str(exc)[:300], ok=False)

    if claim_daily("housekeeping", "03:00"):
        conn = connect()
        try:
            audit.prune(conn)
            shares.purge_expired(conn)
        finally:
            conn.close()


def scheduler_loop() -> None:
    while True:
        for tick in (scheduler_tick, idle_tick, maintenance_tick):
            try:
                tick()
            except Exception:  # noqa: BLE001
                pass
        time.sleep(20)

File: dashboard-ts/src/scheduler.ts
/**
 * The background loop: due schedules, idle machines, nightly housekeeping.
 *
 * The conditional-UPDATE claiming below is no longer strictly required — one
 * Bun process replaces the several gunicorn workers that each ran this loop —
 * but it is kept because it is also what makes it safe to point a second
 * dashboard at the same database, and because losing it would be invisible
 * until the day someone does.
 */

import * as audit from "./audit";
import * as auth from "./auth";
import * as backups from "./backups";
import { providerFor } from "./providers";
import * as shares from "./shares";
import { all, one, run } from "./db";
import { listComputers, type Computer } from "./machines";
import { envInt, nowIso, parseIso } from "./settings";
import { dispatchTask } from "./tasks";

export function computeNextRun(
  kind: string,
  everyMinutes: number | null,
  atTime: string | null,
  after?: Date,
): string {
  const now = after ?? new Date();
  if (kind === "interval") {
    return nowIso(new Date(now.getTime() + (everyMinutes || 60) * 60_000));
  }
  const [hh, mm] = (atTime || "09:00").split(":");
  const next = new Date(now);
  next.setUTCHours(parseInt(hh, 10), parseInt(mm, 10), 0, 0);
  if (next.getTime() <= now.getTime()) next.setUTCDate(next.getUTCDate() + 1);
  return nowIso(next);
}

interface ScheduleRow {
  id: number;
  desktop: string;
  description: string;
  kind: string;
  every_minutes: number | null;
  at_time: string | null;
  next_run_at: string;
}

/**
 * Dispatch every schedule that has come due.
 *
 * Claiming is a conditional UPDATE on next_run_at: only the caller whose
 * UPDATE matches the row it read gets to fire, so a schedule never
 * double-dispatches.
 */
export function schedulerTick(): void {
  const now = new Date();
  const due = all<ScheduleRow>(
    "SELECT * FROM schedules WHERE enabled = 1 AND next_run_at <= ?",
    nowIso(now),
  );

  for (const row of due) {
    const next = computeNextRun(row.kind, row.every_minutes, row.at_time, now);
    const claimed = run(
      "UPDATE schedules SET next_run_at = ?, last_run_at = ?, run_count = run_count + 1 " +
        "WHERE id = ? AND next_run_at = ?",
      next,
      nowIso(now),
      row.id,
      row.next_run_at,
    );
    if (claimed.changes !== 1) continue; // someone else got it first
    try {
      dispatchTask(row.desktop, row.description);
    } catch {
      /* a schedule that can't dispatch must not stop the loop */
    }
  }
}

/**
 * Put machines nobody is watching to sleep.
 *
 * Two things count as "in use": a browser with the screen open (an established
 * connection to websockify inside the desktop) and a task that is pending or
 * running. Anything else has been idle since last_active_at.
 *
 * Deliberately off by default. Sleeping frees the machine's memory but ends
 * its X session, so a surprise suspend costs someone their open windows.
 * Machines flagged no_suspend are always skipped.
 */
export async function idleTick(): Promise<void> {
  const minutes = envInt("DESKSWARM_IDLE_SUSPEND_MINUTES", 0);
  if (minutes <= 0) return;
  const cutoff = Date.now() - minutes * 60_000;
  const busy = new Set(
    all<{ desktop: string }>(
      "SELECT DISTINCT desktop FROM tasks WHERE status IN ('PENDING','RUNNING')",
    ).map((r) => r.desktop),
  );
  const rows = all<Computer>("SELECT * FROM computers WHERE no_suspend = 0");

  const touch = (id: number) =>
    run("UPDATE computers SET last_active_at = ? WHERE id = ?", nowIso(), id);

  for (const comp of rows) {
    if (busy.has(comp.name)) {
      touch(comp.id);
      continue;
    }
    const backend = providerFor(comp);
    const watchers = await backend.vncWatchers(comp.slug);
    if (watchers === null) continue; // already asleep, or unreachable
    if (watchers > 0) {
      touch(comp.id);
      continue;
    }
    const last = parseIso(comp.last_active_at);
    if (!last) {
      // Never seen active — start the clock now rather than suspending a
      // machine the moment the feature is switched on.
      touch(comp.id);
      continue;
    }
    if (last.getTime() > cutoff) continue;
    try {
      await backend.suspendComputer(comp.slug);
    } catch {
      /* a machine that won't stop is not a reason to skip the rest */
    }
  }
}

/**
 * True at most once per UTC day, for whichever caller gets there first.
 *
 * Same conditional-UPDATE trick the task scheduler uses: a fleet backup that
 * fires once per worker would be three times the disk and three times the wall
 * clock.
 */
export function claimDaily(key: string, atTime: string): boolean {
  const parts = (atTime || "").split(":");
  const hh = parseInt(parts[0], 10);
  const mm = parseInt(parts[1], 10);
  if (!Number.isFinite(hh) || !Number.isFinite(mm)) return false;

  const now = new Date();
  if (
    now.getUTCHours() < hh ||
    (now.getUTCHours() === hh && now.getUTCMinutes() < mm)
  ) {
    return false;
  }
  const today = now.toISOString().slice(0, 10);

  const row = one<{ value: string }>("SELECT value FROM meta WHERE key = ?", key);
  if (row && row.value >= today) return false;
  const claimed = row
    ? run("UPDATE meta SET value = ? WHERE key = ? AND value = ?", today, key, row.value)
    : run("INSERT OR IGNORE INTO meta (key, value) VALUES (?, ?)", key, today);
  return claimed.changes === 1;
}

/** Daily housekeeping: back the fleet up, then trim what has aged out. */
export async function maintenanceTick(): Promise<void> {
  if (backups.DAILY_AT && claimDaily("backup_daily", backups.DAILY_AT)) {
    for (const comp of listComputers()) {
      try {
        const meta = await backups.create(comp.slug);
        audit.record("backup (scheduled)", {
          target: comp.name,
          detail: `${meta.name} (${meta.bytes} bytes)`,
        });
      } catch (err: any) {
        audit.record("backup (scheduled)", {
          target: comp.name,
          detail: String(err?.message ?? err).slice(0, 300),
          ok: false,
        });
      }
    }
  }

  if (claimDaily("housekeeping", "03:00")) {
    audit.prune();
    auth.purgeSessions();
    shares.purgeExpired();
  }
}

let timer: ReturnType<typeof setTimeout> | null = null;

export function startScheduler(): void {
  const loop = async () => {
    for (const tick of [schedulerTick, idleTick, maintenanceTick]) {
      try {
        await tick();
      } catch {
        /* one failing tick must not stop the loop */
      }
    }
    timer = setTimeout(loop, 20_000);
  };
  void loop();
}

export function stopScheduler(): void {
  if (timer) clearTimeout(timer);
  timer = null;
}
